File: src/main/java/shopping/wishlist/controller/WishlistController.java
package shopping.wishlist.controller;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.web.PageableDefault;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import shopping.common.PageResponse;
import shopping.product.controller.dto.ProductResponse;
import shopping.wishlist.controller.dto.WishlistRequest;
import shopping.wishlist.controller.dto.WishlistResponse;
import shopping.wishlist.domain.Wishlist;
import shopping.wishlist.service.WishlistService;

import java.util.List;

@RestController
public class WishlistController {
    private final WishlistService service;

    public WishlistController(WishlistService service) {
        this.service = service;
    }

    @PostMapping("/wishlist")
    public ResponseEntity<Void> create(@RequestAttribute("memberId") Long memberId,
            @RequestBody WishlistRequest request) {
        service.addWishlist(new Wishlist(memberId, request.productId()));
        return ResponseEntity.status(HttpStatus.CREATED).build();
    }

    @GetMapping("/wishlist")
    public ResponseEntity<PageResponse<WishlistResponse>> getWish(
            @RequestAttribute("memberId") Long memberId,
            @PageableDefault(size = 20) Pageable pageable) {
        Page<WishlistResponse> page = service.findWishlistItems(memberId, pageable)
                .map(WishlistResponse::from);
        return ResponseEntity.ok(PageResponse.of(page));
    }

    @DeleteMapping("/wishlist/{id}")
    public ResponseEntity<Void> deleteWish(@RequestAttribute("memberId") Long memberId,
                                           @PathVariable("id") Long id) {
        service.deleteWishlistItem(memberId, id);
        return ResponseEntity.noContent().build();
    }
}

File: src/main/kotlin/shopping/wishlist/controller/WishlistController.kt
package shopping.wishlist.controller

import jakarta.validation.Valid
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.web.PageableDefault
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import shopping.common.api.ApiResponse
import shopping.common.domain.CurrentMember
import shopping.wishlist.application.AddWishlistRequest
import shopping.wishlist.application.AddWishlistResponse
import shopping.wishlist.application.WishlistProductDto
import shopping.wishlist.application.WishlistQueryService
import shopping.wishlist.application.WishlistService

@RestController
@RequestMapping("/wishlists")
class WishlistController(
    private val wishlistService: WishlistService,
    private val wishlistQueryService: WishlistQueryService,
) {
    @PostMapping
    fun addWishlist(
        @Valid @RequestBody request: AddWishlistRequest,
        currentMember: CurrentMember,
    ): ApiResponse<AddWishlistResponse> {
        val response = wishlistService.addWishlist(request = request, currentMember = currentMember)

        return ApiResponse.success(response)
    }

    @GetMapping
    fun getWishlists(
        currentMember: CurrentMember,
        @PageableDefault pageable: Pageable,
    ): ApiResponse<Page<WishlistProductDto>> {
        val response = wishlistQueryService.findAllByUserId(currentMember = currentMember, pageable = pageable)

        return ApiResponse.success(response)
    }

    @DeleteMapping("/{productId}")
    fun deleteWishlist(
        currentMember: CurrentMember,
        @PathVariable("productId") productId: Long,
    ): ResponseEntity<ApiResponse<String>> {
        wishlistService.deleteWishlist(productId = productId, currentMember = currentMember)

        return ResponseEntity.noContent().build()
    }
}
